### PR TITLE
Node register arrow ipc buffer fix

### PR DESCRIPTION
**Pull request by samansmink**
Switch from using the replacement scans to tmp views for registering arrow ipc buffers.

### CODE REVIEW

File: tools/nodejs/src/connection.cpp [hunk 1]
```diff
@@ -506,4 +421,88 @@ Napi::Value Connection::Exec(const Napi::CallbackInfo &info) {
...
+
+	for (uint64_t ipc_idx = 0; ipc_idx < array.Length(); ipc_idx++) {
+		Napi::Value v = array[ipc_idx];
+		if (!v.IsObject()) {
+			Napi::TypeError::New(env, "Incorrect params").ThrowAsJavaScriptException();
+			return env.Null();
+		}
+		Napi::Uint8Array arr = v.As<Napi::Uint8Array>();
+		auto raw_ptr = reinterpret_cast<uint64_t>(arr.ArrayBuffer().Data());
+		auto length = (uint64_t)arr.ElementLength();
```
Comment (Mause) [inline tools/nodejs/src/connection.cpp:463]: How are we ensuring that the array doesn't get GC'd ?

File: tools/nodejs/src/connection.cpp [hunk 2]
```diff
@@ -506,4 +421,88 @@ Napi::Value Connection::Exec(const Napi::CallbackInfo &info) {
...
+	}
+
+	array_references[name] = Napi::Persistent(array);
+
+	std::string arrow_scan_function = "scan_arrow_ipc([";
+
+	for (uint64_t ipc_idx = 0; ipc_idx < array.Length(); ipc_idx++) {
+		Napi::Value v = array[ipc_idx];
+		if (!v.IsObject()) {
+			Napi::TypeError::New(env, "Incorrect params").ThrowAsJavaScriptException();
```
Comment (Mause) [inline tools/nodejs/src/connection.cpp:458]: Can we add the index to the error message?

File: tools/nodejs/src/connection.cpp [hunk 3]
```diff
@@ -506,4 +421,88 @@ Napi::Value Connection::Exec(const Napi::CallbackInfo &info) {
...
+	database_ref->Schedule(info.Env(), duckdb::make_unique<ExecTask>(*this, final_query, callback));
+
+	return Value();
+}
+
+Napi::Value Connection::UnRegisterBuffer(const Napi::CallbackInfo &info) {
+	auto env = info.Env();
+
+	if (info.Length() < 1 || !info[0].IsString()) {
+		Napi::TypeError::New(env, "Holding it wrong").ThrowAsJavaScriptException();
```
Comment (Mause) [inline tools/nodejs/src/connection.cpp:485]: Please use a nicer error message

**Review by Mause — CHANGES_REQUESTED** [review]

File: tools/nodejs/src/connection.cpp (hunk 1 above)
Comment (samansmink) [reply tools/nodejs/src/connection.cpp:463]: We store a persistent reference to the array containing the Uint8Arrays that will be kept in the connection object

Comment (Mause) [reply tools/nodejs/src/connection.cpp:463]: As far as I can see, you're just passing the pointer value into to_string, no? Where's the code to keep the reference so the array isn't GC'd?

Comment (samansmink) [reply tools/nodejs/src/connection.cpp:463]: @Mause ah sorry i missed this message. In line 451 we make a reference to the whole array using Napi::Persistent, so this should keep them from being gc'd?

File: tools/nodejs/src/connection.cpp [hunk 4]
```diff
@@ -506,4 +421,90 @@ Napi::Value Connection::Exec(const Napi::CallbackInfo &info) {
...
+		if (!v.IsObject()) {
+			Napi::TypeError::New(env, "Parameter 2 contains unexpected type at index " + std::to_string(ipc_idx))
+			    .ThrowAsJavaScriptException();
+			return env.Null();
+		}
+		Napi::Uint8Array arr = v.As<Napi::Uint8Array>();
+		auto raw_ptr = reinterpret_cast<uint64_t>(arr.ArrayBuffer().Data());
+		auto length = (uint64_t)arr.ElementLength();
+
+		arrow_scan_function += "{'ptr': " + std::to_string(raw_ptr) + ", 'size': " + std::to_string(length) + "},";
```
Comment (hannes) [inline tools/nodejs/src/connection.cpp:467]: I would recommend to use the relational api for this, e.g. 
```C++
		conn->conn->TableFunction("r_dataframe_scan", {Value::POINTER((uintptr_t)value.data())}, parameter_map)
		    ->CreateView(name, overwrite, true);
```

Comment (samansmink) [reply tools/nodejs/src/connection.cpp:467]: Will rewrite this in next pr, can merge for now so we get this working again 👍

**Pull request merged**